Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,539 @@
+
+/-!
+# Properties of Shannon q-ary entropy and binary entropy functions
+
+The [binary entropy function](https://en.wikipedia.org/wiki/Binary_entropy_function)
+`binaryEntropy p := - p * log p - (1 - p) * log (1 - p)`
+is the Shannon entropy of a Bernoulli random variable with success probability `p`.
+
+More generally, the q-ary entropy function is the Shannon entropy of the random variable
+with possible outcomes `{1, ..., q}`, where outcome `1` has probability `(1 - p)`
+and all other outcomes are equally likely.
+
+`qaryEntropy (q : ℕ) (p : ℝ) := p * log (q - 1) - p * log p - (1 - p) * log (1 - p)`
+
+This file assumes that entropy is measured in Nats, hence the use of natural logarithms.
+Most lemmas are also valid using a logarithm in a different base.
+
+## Tags
+
+entropy, Shannon, binary
+-/
+
+open Real
+
+/-- Shannon q-ary Entropy function (measured in Nats, i.e., using natural logs).
+
+It's the Shannon entropy of a random variable with possible outcomes {1, ..., q}
+where outcome `1` has probability `(1 - p)` and all other outcomes are equally likely.
+
+Usual domain of definition is p ∈ [0,1], i.e., input is a probability.
+
+This is a generalization of the binary entropy function `binaryEntropy`. -/
+noncomputable def qaryEntropy (q : ℕ) (p : ℝ) : ℝ :=
+  p * log (q - 1) - p * log p - (1 - p) * log (1 - p)
+
+/-- The [binary entropy function](https://en.wikipedia.org/wiki/Binary_entropy_function)
+`binaryEntropy p := - p * log p - (1-p) * log (1 - p)`
+is the Shannon entropy of a Bernoulli random variable with success probability `p`. -/
+noncomputable def binaryEntropy := qaryEntropy 2
+
+lemma binaryEntropy_eq : binaryEntropy = (fun p => -p * log p - (1 - p) * log (1 - p)) := by
+  have : (2 : ℝ) - 1 = 1 := by norm_num
+  ext
+  simp [binaryEntropy, qaryEntropy, this]
+
+lemma binaryEntropy_eq' {p : ℝ} : binaryEntropy p = -p * log p - (1 - p) * log (1 - p) := by
+  rw [binaryEntropy_eq]
+
+@[simp] lemma qaryEntropy_zero {q : ℕ} : qaryEntropy q 0 = 0 := by
+  simp only [qaryEntropy, zero_mul, log_zero, mul_zero, sub_self, sub_zero, log_one]
+
+@[simp] lemma binaryEntropy_one : binaryEntropy 1 = 0 := by
+  simp only [binaryEntropy_eq', log_one, mul_zero, sub_self, log_zero]
+
+@[simp] lemma qaryEntropy_one {q : ℕ} : qaryEntropy q 1 = log (q - 1) := by
+  unfold qaryEntropy
+  simp only [log_one, mul_zero, sub_self, log_zero, one_mul, sub_zero]
+
+@[simp] lemma binaryEntropy_onehalf : binaryEntropy 2⁻¹ = log 2 := by
+  simp only [binaryEntropy_eq', show (1 : ℝ) - 2⁻¹ = 2⁻¹ by norm_num, log_inv]
+  field_simp
+
+/-- `binaryEntropy` is symmetric about 1/2. -/
+@[simp] lemma binaryEntropy_one_sub (p : ℝ) :
+    binaryEntropy (1 - p) = binaryEntropy p := by
+  simp only [binaryEntropy_eq', neg_sub, sub_sub_cancel, neg_mul]
+  ring
+
+/-- `binaryEntropy` is symmetric about 1/2. -/
+lemma binaryEntropy_add_onehalf (p : ℝ) :
+    binaryEntropy (2⁻¹ + p) = binaryEntropy (2⁻¹ - p) := by
```

Comment:
```suggestion
lemma binaryEntropy_two_inv_add (p : ℝ) :
    binaryEntropy (2⁻¹ + p) = binaryEntropy (2⁻¹ - p) := by
```
I searched around the docs a bit, and it seems like `2⁻¹` is usually referred to as `two_inv` in lemma names, `1 / 2` as `one_half` (and `half` by itself is the operation of dividing by two). Let's just go with that.